static/js/subs.js

```diff
@@ -727,6 +727,50 @@ $(function () {
         update_announce_stream_state();
     });
 
+    // Search People
+    $(document).on('input', '.add-user-list-filter', function (e) {
+        var users = people.get_rest_of_realm();
+
+        var user_list = $(".add-user-list-filter");
+        if (user_list === 0) {
+            return;
+        }
+        var search_term = user_list.expectOne().val().trim();
+        var search_terms = search_term.toLowerCase().split(",");
+        search_terms = _.map(search_terms, function (s) {
+            return s.trim();
+        });
+
+        var filtered_users = _.filter(users, function (user) {
+            var person = people.get_by_email(user.email);
+            if (!person || !person.full_name) {
+                return false;
+            }
+            var names = person.full_name.toLowerCase().split(/\s+/);
+            names = _.map(names, function (s) {
+                return s.trim();
+            });
+            return _.any(search_terms, function (search_term) {
+                return _.any(names, function (name) {
+                    return name.indexOf(search_term) === 0;
+                });
+            });
+        });
+
+        _.each(users, function (user) {
```

Comment:
That method results in an O(n^2) algorithm, because you loop through each user in `_.each` and then for each user you have to loop through `filtered_users` in `_.where`. You avoid this extra O(n) factor if you make `filtered_users` an object since key lookup is O(1).

Broadly speaking, a linear algorithm is typically fine while a quadratic algorithm is not. For instance, if there are 1000 users, then the linear algorithm is O(1000) while the quadratic algorithm is O(1,000,000).